### PR TITLE
Repair other units

### PR DESCRIPTION
**Pull request by tbeddy**
This allows specific unit types to repair other units. Currently only the medic is able to do so. Images have been added for the medic unit-type, along with an image to highlight the border of targeted friendly units in green. Bugs include medics being unable to repair others after moving once and inconsistent tile-masking.
Update: I've fixed the first bug. The tile-masking is improved, though probably not ideal.

### CODE REVIEW

File: src/cljs/zetawar/events/game.cljs [hunk 1]
```diff
@@ -23,6 +23,13 @@
          :dispatch [[:zetawar.events.ui/hide-copy-link]]
          :notify   [[:zetawar.players/apply-action :faction.color/all action]]})
 
+      :action.type/repair-other-unit
```
Comment (djwhitt) [inline src/cljs/zetawar/events/game.cljs]: This is the same as the default case, isn't it? Could it just be eliminated and fall through to the default?

File: src/cljs/zetawar/data.cljs [hunk 2]
```diff
@@ -133,6 +136,7 @@
                 :cost 75
                 :movement 9
                 :can-capture true
+                :can-repair false
```
Comment (djwhitt) [inline src/cljs/zetawar/data.cljs]: It complicates things a bit, but I think this should be a set of armor classes rather than a boolean. That way we can have engineers that can repair armored units, etc.

File: src/cljs/zetawar/data.cljs [hunk 3]
```diff
@@ -31,6 +31,7 @@
     :build-unit-button "Build"
     :attack-unit-button "Attack"
     :repair-unit-button "Repair"
+    :repair-other-button "Repair Other"
```
Comment (djwhitt) [inline src/cljs/zetawar/data.cljs]: Although we need to distinguish between self repair and repairing other units at the action level, I think it's ok for the buttons to have the same label. The context makes the type of repair clear.

File: src/cljs/zetawar/events/ui.cljs [hunk 4]
```diff
@@ -210,6 +213,21 @@
                   :action/r r}]
                 [::clear-selection]]}))
 
+(defmethod router/handle-event ::repair-targeted
+  [{:as handler-ctx :keys [db]} _]
+  (let [game (app/current-game db)
+        cur-faction-color (game/current-faction-color game)
+        [repairer-q repairer-r] (app/selected-hex db)
+        [wounded-q  wounded-r]  (app/targeted-hex db)]
```
Comment (djwhitt) [inline src/cljs/zetawar/events/ui.cljs]: Just a style comment, but I'd remove the extra whitespace here. I *mostly* don't try to align things in let statements. I like the way it looks, but it's just too hard to be consistent about it imo.

File: src/cljs/zetawar/subs.cljs [hunk 5]
```diff
@@ -431,11 +476,24 @@
...
+  (when-let [[selected-q selected-r] @(selected-hex conn)]
+    @(in-range-of-friend-at? conn selected-q selected-r q r)))
+
 (deftrack selected-can-attack-targeted? [conn]
   (when-let [[q r] @(targeted-hex conn)]
     (and @(selected-can-attack? conn)
          @(enemy-in-range-of-selected? conn q r))))
 
+(deftrack selected-can-repair-targeted? [conn]
+  ;;Is the selected able to repair others?
```
Comment (djwhitt) [inline src/cljs/zetawar/subs.cljs]: The code is pretty clear. I'm don't think these comments are necessary.

File: src/cljs/zetawar/views.cljs [hunk 6]
```diff
@@ -74,9 +81,20 @@
                    (not @(subs/unit-can-act? conn q r)))
 
               ;; Unit selected and tile is a valid attack or move target
+              ;; However, selected cannot repair others
+              (and @(subs/unit-selected? conn)
```
Comment (djwhitt) [inline src/cljs/zetawar/views.cljs]: I think you can combine the 'and's on lines 85 and 93 by adding '@(subs/selected-can-repair-other? conn)' into the 'not' + 'and' currently on line 96 and moving into the higher level and starting on line 85 (hope that makes sense + is correct :smile:).

File: src/cljs/zetawar/subs.cljs [hunk 7]
```diff
@@ -431,11 +476,24 @@
   (when-let [[selected-q selected-r] @(selected-hex conn)]
     @(in-range-of-enemy-at? conn selected-q selected-r q r)))
 
+(deftrack friend-in-range-of-selected? [conn q r]
```
Comment (djwhitt) [inline src/cljs/zetawar/subs.cljs:493]: I think it might simplify logic elsewhere to either turn this into 'repairable-friend-in-range-of-selected' (also checks if friends are repairable) or add that as a second track.

**Review by djwhitt — CHANGES_REQUESTED** [review]
Initial review feedback

File: src/cljs/zetawar/data.cljs (hunk 3 above)
Comment (tbeddy) [reply src/cljs/zetawar/data.cljs]: That sounds reasonable.

File: src/cljs/zetawar/events/ui.cljs (hunk 4 above)
Comment (tbeddy) [reply src/cljs/zetawar/events/ui.cljs]: Understandable. I tried to find a pair of words with the same length to match "attacker/defender," but couldn't decide on anything.

File: src/cljs/zetawar/subs.cljs (hunk 7 above)
Comment (tbeddy) [reply src/cljs/zetawar/subs.cljs:493]: That sounds good.

File: src/cljs/zetawar/subs.cljs (hunk 5 above)
Comment (tbeddy) [reply src/cljs/zetawar/subs.cljs]: Yeah, the comments were only there for me to think it through.

File: src/cljs/zetawar/views.cljs (hunk 6 above)
Comment (tbeddy) [reply src/cljs/zetawar/views.cljs]: That indeed makes sense and is correct.

File: src/cljs/zetawar/data.cljs (hunk 2 above)
Comment (tbeddy) [reply src/cljs/zetawar/data.cljs]: Okay. Should medics only be able to repair/heal other personnel?

Comment (tbeddy) [reply src/cljs/zetawar/data.cljs]: And what would you like to call each of the armor classes? :personnel for medics, :armored for engineers, and :none for everyone else?

Comment (djwhitt) [reply src/cljs/zetawar/data.cljs]: The armor classes (I guess I actually called them armor-types...) should be the same as those used for zoc and attack strength. Engineers should be able to repair 'armored' units and medics should be able to repair 'personnel'. All other units get an empty set.

Comment (tbeddy) [reply src/cljs/zetawar/data.cljs]: It works, but the work I did in subs.cljs might seem a little janky. There's probably a better way to handle the selected and targeted values, but I'm not sure what that is.

**Comment by djwhitt** [comment]
The use of sets for :can-repair in data.cljs is good, but they needed to be loaded into the DataScript DB a little differently. Each item in the set should result in a datom attached to the unit type entity in the DB. The closest analog for how this should work is :unit-type/zoc-armor-types.

**Comment by tbeddy** [comment]
I'm not familiar with DataScript, so I'm having some trouble with this. Should I be altering the schema as defined in db.cljs?

**Comment by djwhitt** [comment]
Yep, attributes with cardinality > 1 need schema entries. It should look something like this: https://github.com/Zetawar/zetawar/blob/5672efd67f8b8a2a06e5e67911d83c32cc039b6f/src/cljs/zetawar/db.cljs#L56

**Comment by tbeddy** [comment]
Is that last commit closer?

**Comment by djwhitt** [comment]
Hmm... that is odd. I'll take a closer look.

**Comment by tbeddy** [comment]
On a different note, are you okay with two of the repair functions I added in game.cljs (can-be-repaired? and compatible-armor-types-for-repair?) not following the check-next-state model you use for can-repair?, can-attack?, etc.?

**Comment by djwhitt** [comment]
You mean should a unit be able to repair another unit regardless of what state the other unit is in? If that's the question, I'd say yes.

**Comment by tbeddy** [comment]
I mean should a unit have to check its state at all to be able to repair another unit. 

can-be-repaired? is currently... 
`(defn can-be-repaired? [db game unit] 
     (< (:unit/count unit) (:game/max-count-per-unit game)))`

(which doesn't use the database at all)

...instead of something like...
`(defn can-be-repaired? [db game unit]
    (try (check-can-be-repaired db game unit)
    true
    (catch :default ex
      false)))`

...which is how most of the surrounding functions have been designed.

**Comment by djwhitt** [comment]
Ah, I see what you mean. It doesn't really need to check the state, but it probably should work the same way as 'check-in-range' where there is both a 'check' version that throws an exception and a '?' version that returns a value. That way you can call the 'check' version in the 'repair-other-tx' as an assertion.

**Comment by tbeddy** [comment]
I adapted those functions. Let me know if they look alright and anything else you'd like worked on when you have the chance.

**Comment by djwhitt** [comment]
I'm traveling now, so it might be a few days before I can take a look. But I'll check it out as soon as I get a chance.

File: src/cljs/zetawar/subs.cljs [hunk 8]
```diff
@@ -456,8 +455,9 @@
 
 (deftrack compatible-armor-types-for-repair? [conn targeted-q targeted-r]
   (when-let [[selected-q selected-r] @(selected-hex conn)]
-    (game/compatible-armor-types-for-repair? @conn @(game conn) @(unit-at conn selected-q selected-r)
-                                                                @(unit-at conn targeted-q targeted-r))))
+    (game/compatible-armor-types-for-repair? @conn @(game conn)
```
Comment (djwhitt) [inline src/cljs/zetawar/subs.cljs]: What's the reason for this change? It should work either way as long as the needed properties are pulled by the unit-at sub, right?

Comment (tbeddy) [reply src/cljs/zetawar/subs.cljs]: There was a problem when doing the checks on circleci that seemed to stem from using :unit-type/armor-type in the subs/unit-at function. Adapting subs_test.cljs to accommodate this wasn't working, so I used game/unit-at instead. This may just be a result of my inexperience with both git and testing frameworks like circleci, so I can try to fix it.

Comment (tbeddy) [reply src/cljs/zetawar/subs.cljs]: I made the changes and the tests passed.
Should I move the compatible-armor-types-for-repair? track out of the "Selection and target" section and into the "Units" section? It uses the selected-hex track and calls a function from game.cljs, so it's kind of in-between. Maybe break it up into two tracks?

Comment (djwhitt) [reply src/cljs/zetawar/subs.cljs]: Hmm... maybe. I'm considering just merging it as is, doing a bit of cleanup myself, and then getting you to review that. The current state is pretty good and all the back and forth in comments can get a bit tedious. Master only gets deployed to dev.zetawar.com, so if something is still a little off it's not going to mess up anyone's games.

Comment (tbeddy) [reply src/cljs/zetawar/subs.cljs]: Cool, that sounds like a good plan.

**Review by djwhitt — APPROVED** [review]

**Comment by djwhitt** [comment]
Merged. Thanks for all your work on this! I'll get started on a review/cleanup branch soon and ping you when it's ready.

**Comment by tbeddy** [comment]
Cool, thanks for all your help.
